### ISSUE TITLE
Need SPARQL query to build relationship graph from a selected term.

### ISSUE THREAD

**Issue by paulpilone** [iodepo/OceanBestPractices#118]
The user is able to select an existing term and the application should display the relationships (defined by IODEPO) for that term. We need SPARQL queries that can build that relationship for a given term. I assume we'll need queries specific to each ontology (or at least groups of ontologies). I can offer the current query as an example but it's so complex (and wrong) that I'm not sure it'll be of much help.

This is to address OBP-256: As the owner, I want to see expected semantic neighborhoods with full and accurate relationships.

**Comment by pbuttigieg** [comment]
The semantic neighbourhood would be any OWL class or SKOS concept linked to the target term with a meaningful relationship.

To make this easier for devs/maintainers and less SPARQL-focused, I suggest we do the following: 

### Step 1

Get all the `subject - predicate - object` triples from the target ontology / thesaurus with some filtering as follows...

For an OWL ontology, we want to filter away things like annotation properties and datatype properties and focus on object properties, so something like this could work:

```sparql
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT ?s ?slabel ?p ?plabel ?o ?olabel 

FROM [graphName]

WHERE
 {
  ?s  ?p ?o.
  
  values ?ptype { owl:ObjectProperty } 
  ?p a ?ptype.

  ?s rdfs:label ?slabel.
  ?p rdfs:label ?plabel.
  ?o rdfs:label  ?olabel.
}
ORDER BY ASC(?s)

```
That should get you a result with all triples that have object properties as predicates. 

For SKOS thesauri (e.g. the NERC stuff) you'd switch out the ObjectProperty with the SKOS semantic relations like broader, narrower, or close matches (see [here](https://www.w3.org/TR/skos-reference/#semantic-relations)). 

For things like WoRMS, this would depend on how they are distributed, and I'd need to see what we're working with. xref #116  

## Step 2

With that result set in hand, you can write a generic python script (or similar) that takes a tag, grabs its IRI, finds and extracts s-p-o triples where the tag matches the ?s, and then builds your neighbourhood (in RDF).

To save repeated computing, it would probably be best to pre-compute / pre-extract all semantic neighbourhoods for each ?s you get from Step 1 (basically splitting up the result set). Then you can map all tags in the docs to their pre-computed semantic neighbourhoods quite easily.

**Comment by paulpilone** [comment]
I don't believe the above query works for obo ontologies. For example running that query against ENVO only results in ~36 labels and I'm pretty sure it has more than that.

Also tried the following query against L05 (SKOS) and it returned 0 results:

```
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT ?s ?slabel

FROM <http://vocab.nerc.ac.uk/collection/L05/current/>

WHERE
 {
  ?s  ?p ?o.

  values ?ptype { skos:semanticRelation skos:broader skos:narrower skos:related skos:broaderTransitive skos:narrowerTransitive }
  ?p a ?ptype.

  ?s rdfs:label ?slabel.
}
ORDER BY ASC(?s)
```

**Comment by paulpilone** [comment]
@pbuttigieg Could you provide a full example of your solution for one of our ontologies (ENVO is fine). When I use the query you provided I get back a pretty small list of triples. To me, it doesn't feel like I'm finding everything I need to find.

Assuming that list is correct (or we update the query to get the correct list) I'm not sure I fully understand step 2 and how to build out the relationship for each term so that we can produce it later on.

Also, could you provide guidance on the query I tested against a SKOS vocabulary.
